cub/test/catch2_test_device_merge.cu

```diff
@@ -0,0 +1,462 @@
+  c2h::host_vector<Key> reference_keys_h(size1 + size2);
+  c2h::host_vector<Value> reference_values_h(size1 + size2);
+  {
+    c2h::host_vector<Key> keys1_h     = keys1_d;
+    c2h::host_vector<Value> values1_h = values1_d;
+    c2h::host_vector<Key> keys2_h     = keys2_d;
+    c2h::host_vector<Value> values2_h = values2_d;
+    using value_t                     = typename decltype(zip(keys1_h.begin(), values1_h.begin()))::value_type;
+    std::merge(zip(keys1_h.begin(), values1_h.begin()),
+               zip(keys1_h.end(), values1_h.end()),
+               zip(keys2_h.begin(), values2_h.begin()),
+               zip(keys2_h.end(), values2_h.end()),
+               zip(reference_keys_h.begin(), reference_values_h.begin()),
+               [&](const value_t& a, const value_t& b) {
+                 return compare_op(thrust::get<0>(a), thrust::get<0>(b));
+               });
+  }
+
+  // FIXME(bgruber): comparing std::vectors (slower than thrust vectors) but compiles a lot faster
+  CHECK((detail::to_vec(reference_keys_h) == detail::to_vec(c2h::host_vector<Key>(result_keys_d))));
+  CHECK((detail::to_vec(reference_values_h) == detail::to_vec(c2h::host_vector<Value>(result_values_d))));
+}
+
+CUB_TEST("DeviceMerge::MergePairs key types", "[merge][device]", types)
+{
+  using key_t    = c2h::get<0, TestType>;
+  using value_t  = int;
+  using offset_t = int;
+  test_pairs<key_t, value_t, offset_t>();
+}
+
+// TODO(bgruber): fine tune the type sizes again to hit the fallback and the vsmem policies
+// CUB_TEST("DeviceMerge::MergePairs large key types", "[merge][device]", large_types)
+// {
+//   using key_t    = c2h::get<0, TestType>;
+//   using value_t  = int;
+//   using offset_t = int;
+//   test_pairs<key_t, value_t, offset_t>();
+// }
+
+CUB_TEST("DeviceMerge::MergePairs value types", "[merge][device]", types)
+{
+  using key_t    = int;
+  using value_t  = c2h::get<0, TestType>;
+  using offset_t = int;
+  test_pairs<key_t, value_t, offset_t>();
+}
+
+CUB_TEST("DeviceMerge::MergePairs offset types", "[merge][device]", offset_types)
+{
+  using key_t    = int;
+  using value_t  = int;
+  using offset_t = c2h::get<0, TestType>;
+  test_pairs<key_t, value_t, offset_t>(3623, 6346, ::cuda::std::less<>{}, merge_pairs_custom_offset_type);
+}
+
+CUB_TEST("DeviceMerge::MergePairs input sizes", "[merge][device]")
+{
+  using key_t      = int;
+  using value_t    = int;
+  using offset_t   = int;
+  const auto size1 = offset_t{GENERATE(0, 1, 23, 123, 3234234)};
+  const auto size2 = offset_t{GENERATE(0, 1, 52, 556, 56767)};
+  test_pairs<key_t, value_t>(size1, size2);
+}
+
+// this test exceeds 4GiB of memory and the range of 32-bit integers
+CUB_TEST("DeviceMerge::MergePairs really large input", "[merge][device]")
+try
+{
+  using key_t     = char;
+  using value_t   = char;
+  const auto size = std::int64_t{1} << GENERATE(30, 31, 32, 33);
+  test_pairs<key_t, value_t>(size, size, ::cuda::std::less<>{}, merge_pairs_custom_offset_type);
```

Comment:
important: this test doesn't work at the moment. It's blocked by https://github.com/NVIDIA/cccl/issues/1951.

Reply:
That's probably only since #1884 was merged, because before `thrust::sort` uses merge sort for primitive types if the comparator is `::cuda::std::less` and not `thrust::less`.